src/spec-common/cliHost.ts

```diff
@@ -87,29 +87,159 @@ function createLocalCLIHostFromExecFunctions(localCwd: string, exec: ExecFunctio
+function skipHeader(headerSize: number, cb: SourceCallback<Buffer>, abort: Abort, data?: Buffer): number {
+	if (abort || data === undefined) {
+		cb(abort);
+		return headerSize;
 	}
 
-	const socket = new net.Socket();
+	if (headerSize === 0) {
+		// Fast path avoiding data buffer manipulation
+		cb(null, data);
+	} else if (data.length > headerSize) {
+		// We need to remove part of the data to forward
+		let removeSize = data.length - headerSize;
+		data.copy(data, 0, removeSize);
+		headerSize = 0;
+		cb(null, data);
+	} else {
+		// We need to remove all forwarded data
+		headerSize = headerSize - data.length;
+		cb(null, Buffer.of());
+	}
+
+	// Return the updated headerSize
+	return headerSize;
+}
+
+// Function to handle the Cygwin/Gpg4win socket filtering
+// These sockets need an handshake before forwarding client and server data
+function handleUnixSocketOnWindows(socket: net.Socket, socketPath: string): Duplex<Buffer, Buffer> {
+	let headerSize = 0;
+	let pendingSourceCallbacks: { abort: Abort; cb: SourceCallback<Buffer> }[] = [];
+	let pendingSinkCalls: Source<Buffer>[] = [];
+	let connectionDuplex: Duplex<Buffer, Buffer> | undefined = undefined;
+
 	(async () => {
 		const buf = await readLocalFile(socketPath);
-		const i = buf.indexOf(0xa);
-		const port = parseInt(buf.slice(0, i).toString(), 10);
-		const guid = buf.slice(i + 1);
+		const str = buf.toString();
+
+		// Try to parse cygwin socket data
+		const cygwinSocketParameters = str.match(/!<socket >(\d+)( s)? ((([A-Fa-f0-9]{2}){4}-?){4})/);
+
+		let port: number;
+		let handshake: Buffer;
+
+		if (cygwinSocketParameters !== null) {
+			// Cygwin / MSYS / Git Bash unix socket on Windows
+			const portStr = cygwinSocketParameters[1];
+			const guidStr = cygwinSocketParameters[3];
+			port = parseInt(portStr, 10);
+			const guid = cygwinUnixSocketCookieToBuffer(guidStr);
+
+			let identifierData = Buffer.alloc(12);
+			identifierData.writeUInt32LE(process.pid, 0);
+
+			handshake = Buffer.concat([guid, identifierData]);
+
+			// Recv header size = GUID (16 bytes) + identifiers (3 * 4 bytes)
+			headerSize = 16 + 3 * 4;
+		} else {
+			// Gpg4Win unix socket
+			const i = buf.indexOf(0xa);
+			port = parseInt(buf.slice(0, i).toString(), 10);
+			handshake = buf.slice(i + 1);
+
+			// No header will be received from Gpg4Win agent
+			headerSize = 0;
+		}
+
 		socket.connect(port, '127.0.0.1', () => {
-			socket.write(guid, err => {
+			connectionDuplex = toPull.duplex(socket);
+
+			// Call pending source calls, if the pull-stream connection was pull-ed
+			// before we got connected to the ssh-agent/gpg-agent server
+			for (let callback of pendingSourceCallbacks) {
+				(connectionDuplex as Duplex<Buffer, Buffer>).source(callback.abort, function (abort, data) {
+					headerSize = skipHeader(headerSize, callback.cb, abort, data);
+				});
+			}
+
+			// Write handshake data to the ssh-agent/gpg-agent server
+			socket.write(handshake, err => {
```

Comment:
I've checked this sequence is possible, but it doesn't triggered while running 3 parallel busy loops of ssh-add to test the code.
But will do the modification anyway just in case.